proposals/NNNN-global-actor-isolated-types-usability.md

```diff
@@ -0,0 +1,225 @@
+
+```swift
+func test(globallyIsolated: @escaping @MainActor () -> Void) {
+  Task {
+    await globallyIsolated() //okay
+  }
+}
+```
+
+The `globallyIsolated` closure in the above code is global-actor isolated via the `@MainActor`. Thus, it can never operate on the same reference concurrently at the same time, making it safe to be invoked from different isolation domains. This means that for such global-actor-isolated closures and functions, the `@Sendable` attribute is implicit.
+
+#### Non-`Sendable` captures in isolated closures
+
+Under this proposal, globally-isolated closures are allowed to capture non-`Sendable` values:
+
+```swift
+class NonSendable {}
+
+func test() {
+  let ns = NonSendable()
+
+  let closure { @MainActor in
+    print(ns)
+  }
+
+  Task {
+    await closure() // okay
+  }
+}
+```
+
+The above code is data-race safe, since a globally-isolated closure will never operate on the same instance of `NonSendable` concurrently.
+
+Note that under region isolation in SE-0414, capturing a non-`Sendable` value in an actor-isolated closure will transfer the region into the actor, so it is impossible to have concurrent access on non-`Sendable` captures even if the isolated closure is formed outside the actor.
+
+
+### Global actor isolation and inheritance
+
+Subclasses may add global actor isolation when inheriting from a nonisolated, non-`Sendable` superclass. In this case, an implicit conformance to `Sendable` will not be added, and explicitly specifying a `Sendable` conformance is an error:
+
+```swift
+class NonSendable {
+  func test() {}
+}
+
+@MainActor
+class IsolatedSubclass: NonSendable {
+  func trySendableCapture() {
+    Task.detached {
+      self.test() // error: Capture of 'self' with non-sendable type 'IsolatedSubclass' in a `@Sendable` closure
+    }
+  }
+}
+```
+
+Inherited and overridden methods still must respect the isolation of the superclass method:
+
+```swift
+class NonSendable {
+  func test() { ... }
+}
+
+@MainActor
+class IsolatedSubclass: NonSendable {
+  var mutable = 0
+  override func test() {
+    super.test()
+    mutable += 0 // error: Main actor-isolated property 'mutable' can not be referenced from a non-isolated context
+  }
+}
+```
+
+Matching the isolation of the superclass method is necessary because the superclass method implementation may internally rely on the static isolation, such as when hopping back to the isolation after any asynchronous calls, and because there are a variety of ways to call the subclass method that don't preserve its isolation, including:
+
+* Upcasting to the superclass type
+* Erasing to an existential type based on conformances of the superclass type
+* Passing the isolated subclass as a generic argument to a type parameter that requires a conformance implemented by the superclass
+
+## Source compatibility
+
+The introduced changes are additive, so the proposal does not impact source compatibility.
```

Comment:
The global-actor-implies-`@Sendable` part of this proposal is not additive because it changes the interpretation of existing code.  A call to a function which takes a global-actor-isolated-but-non-`@Sendable` function can become ill-formed under this proposal.

Reply:
Yeah, agree